src/sdk/namespace_cache.js

```diff
@@ -136,12 +201,30 @@ class NamespaceCache {
             const cache_params = { ...params, source_stream: cache_stream };
             const cache_promise = this.namespace_nb.upload_object(cache_params, object_sdk);
 
+            // One important caveat is that if the Readable stream emits an error during processing,
+            // the Writable destination is not closed automatically. If an error occurs, it will be
+            // necessary to manually close each stream in order to prevent memory leaks.
+            params.source_stream.on('error', err => {
+                dbg.log0("NamespaceCache.upload_object: error in read source", {params: _.omit(params, 'source_stream'), error: err});
+                hub_stream.destroy();
+                cache_stream.destroy();
+            });
+
             params.source_stream.pipe(hub_stream);
             params.source_stream.pipe(cache_stream);
 
-            const [hub_res, cache_res] = await Promise.all([hub_promise, cache_promise]);
-            assert.strictEqual(hub_res.etag, cache_res.etag);
-            return hub_res;
+            const hub_upload = await Promise.all([ hub_promise, cache_promise ])
+            .then(([hub_res, cache_res]) => {
+                assert.strictEqual(hub_res.etag, cache_res.etag);
+                return hub_res;
+            })
+            .catch(err => {
+                dbg.log0("NamespaceCache.upload_object: error in upload", {params, error: err});
+                // If error is from cache, we should probably ignore and let hub upload continue. Change Promise.all to Promise.allSettled?
+                throw err;
+            });
```

Comment:
This error handling wouldn't really work - you need to use try-catch around await.
Also we need to return just a single result and not the array of two results.
I agree with the comment you wrote and we should change it to allSettled.

See suggested code -

```suggestion
            const [hub_res, cache_res] = await Promise.allSettled([ hub_promise, cache_promise ]);
            const hub_ok = hub_res.status === 'fulfilled';
            const cache_ok = cache_res.status === 'fulfilled';

            if (!hub_ok) {
                dbg.log0("NamespaceCache.upload_object: error in upload", params, hub_res, cache_res);

                // handling the case where cache succeeded and cleanup.
                // We can also just mark the cache object for re-validation
                // to make sure any read will have to re-validate it,
                // but writes (retries of the upload most likely) will be already in the cache
                // and detected by dedup so we don't need to do anything.
                if (cache_ok) {
                    try {
                        await this. _delete_object_from_cache(params, object_sdk);
                    } catch (err) {
                        dbg.error('Failed to delete object from cache after hub failure', err.stack);
                    }
                }

                // fail back to client with the hub reason
                throw hub_res.reason;
            }
            
            if (cache_ok) {
                assert.strictEqual(hub_res.value.etag, cache_res.value.etag);
            } else {
                // on error from cache, we ignore and let hub upload continue
                dbg.log0("NamespaceCache.upload_object: error in cache upload", params, hub_res, cache_res);
            }

            return hub_res.value;
```

Reply:
ok